Add modulation input for DeformConv2D

Licht-T opened the pull request:
This is the implementation of [Modulated Deformable Convolution, a.k.a. DCNv2](https://arxiv.org/abs/1811.11168). This closes #1788.

fmassa reviewed (commented):
Thanks a lot for the PR!

I haven't yet gone in details in the C++ / CUDA parts, but I have a few comments in the python part that could help fix the torchscript tests.

Licht-T commented on torchvision/ops/deform_conv.py:
```diff
@@ -12,6 +12,7 @@
 def deform_conv2d(
     input: Tensor,
     offset: Tensor,
+    mask: Optional[Tensor],
```
Okay! How about C++ API?

fmassa replied:
I would be ok breaking the C++ API -- we don't have as much constraints as with Python nowadays, so on the C++ API this is fine

Licht-T replied:
Fixed! https://github.com/pytorch/vision/pull/2791/commits/2c60f5cc27c9dabac06790a7d4b7c5c3ac084df4

Licht-T commented on torchvision/ops/deform_conv.py:
```diff
@@ -73,10 +84,27 @@ def deform_conv2d(
+            or mask.shape[1] != mask_ch
+            or mask.shape[2] != mask_h
+            or mask.shape[3] != mask_w
+    ):
```
Thanks for your suggestion. I found that I already put the same assertion in C++ code, and removed the Python's one.

Licht-T commented on torchvision/ops/deform_conv.py:
```diff
@@ -42,18 +46,25 @@ def deform_conv2d(
     _assert_has_ops()
     out_channels = weight.shape[0]
+
+    use_mask = mask is not None
```
As mentioned above, I removed the shape assertion in Python code. I also found that it is better to pass `use_mask` flag into C++ code for shape assertions, so I did.

Licht-T commented:
@fmassa Thanks for you review! I changed my code. TorchScript now works well.

Also, I have changed some assertion logics (`use_mask` related) so that it can correctly determine whether it use the modulation or not. I've tested this on CPU forward/backward. Unfortunately, since my GPU environment has broken last night, I have not yet tested this on GPU. I'll fix up my environment in couple of days and test this on GPU.

Licht-T commented:
Hmmm..., the GPU CI test on Windows works fine, but the Linux one isn't. That's weird...

Licht-T commented:
Anyway, I'll check as soon as my GPU environment fixed!

fmassa commented:
Sounds good.

The tests are randomized, so it might be that in some cases they pass and others they fail. But the gradchecks that are erroing on Linux GPU don't seem to be due to numerical error, so there might be a case that has been missed somewhere.

I'll rerun the tests to see if the CI status change.

Licht-T commented:
@fmassa Thanks for your CI re-run, but the situation didn't change. 

My Linux GPU environment is finally back, and my PR passes GPU tests! This is strange thing, and I am still digging into this.

Licht-T commented:
I run `test_ops.py` tests on the AWS g4dn.xlarge instance, and there was no test failure.
```
(base) ubuntu@ip-172-31-29-152:~/vision$ uname -a
Linux ip-172-31-29-152 5.4.0-1025-aws #25~18.04.1-Ubuntu SMP Fri Sep 11 12:03:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
(base) ubuntu@ip-172-31-29-152:~/vision$ nvidia-smi
Sat Oct 17 13:22:19 2020
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.80.02    Driver Version: 450.80.02    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
| N/A   26C    P8     9W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
(base) ubuntu@ip-172-31-29-152:~/vision$ pytest test/test_ops.py
================================= test session starts ==================================
platform linux -- Python 3.8.3, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/ubuntu/vision
plugins: cov-2.10.1
collected 62 items

test/test_ops.py ..............................................................  [100%]

============================ 62 passed in 321.95s (0:05:21) ============================
(base) ubuntu@ip-172-31-29-152:~/vision$
```

Licht-T commented:
I found insufficient resources errors on CI logs. This test failure only seems to be happened on the environment which has small GPU resources, such as CI. 
https://app.circleci.com/pipelines/github/pytorch/vision/4464/workflows/4d7daf2d-7466-460c-a1f7-76ae3495653c/jobs/266953/steps
```
error in compute_grad_offset_and_mask: too many resources requested for launch
```

Licht-T commented:
@fmassa The Linux GPU CI runs on a small instance. On the other hand, the Windows one is on a medium instance. This led to the test result difference between Linux and Windows. We can reduce # of parallelization, but this makes a reduction in the performance. I recommend that we use the medium instance on Linux GPU CI.
https://github.com/pytorch/vision/blob/a9c78f1b017d912fa9583ae6d93bff0a3807c29b/.circleci/config.yml#L442
https://github.com/pytorch/vision/blob/a9c78f1b017d912fa9583ae6d93bff0a3807c29b/.circleci/config.yml#L18

Licht-T commented:
@fmassa I tried the running GPU CI on `gpu.medium` instance, but it did not make sense. Finally, I found that the Linux environment is CUDA Compute Capability (CC) 5.2, but the Windows one is CC 7.5. The Linux instance has the old GPU even if it is "medium" 🤔 .

According to [the official calculator](https://docs.nvidia.com/cuda/cuda-occupancy-calculator/CUDA_Occupancy_Calculator.xls), the CC < 6 has smaller number of registers per block than CC >= 6 and cannot launch the kernel which uses a large number of registers. So I decide to select the proper thread size whether CC is higher than 6 or not.

And now, the PR passed all tests. All-green, except Travis CI!

fmassa reviewed (commented):
thanks a lot for the PR and sorry for the delay in getting back to you, the release was taking most of our time over the past couple of weeks.

This PR looks good to merge, but there are some merge conflicts now, could you look into fixing those?

Licht-T commented:
@fmassa No problem! Now, fixed conflicts and CI is all green except CodeCov!

Also, I made https://github.com/pytorch/vision/pull/2973 since GPU CI was not working well.

fmassa commented on .circleci/config.yml.in:
```diff
@@ -455,21 +455,17 @@ jobs:
-      - restore_cache:
-          {% raw %}
-          keys:
-            - env-v2-linux-{{ arch }}-py<< parameters.python_version >>-{{ checksum ".circleci/unittest/linux/scripts/environment.yml" }}-{{ checksum ".circleci-weekly" }}
```
I believe the recommended way to handle invalid caches on CI is to change the key name, instead of having it be `env-v2` let it be `env-v3` for example.

fmassa commented on .circleci/config.yml.in:
```diff
@@ -455,21 +455,17 @@ jobs:
     resource_class: gpu.small
     environment:
       image_name: "pytorch/manylinux-cuda101"
+      PYTHON_VERSION: << parameters.python_version >>
```
Those changes are also present in #2973, and there is some discussion going on with @mthrok on the best way to solve this.

In order to move forward with this PR, I'll be merging this as is but let's maybe revisit this implementation in a follow-up PR following @mthrok and @seemethere feedback.

fmassa commented on torchvision/ops/deform_conv.py:
```diff
@@ -33,6 +34,9 @@ def deform_conv2d(
         dilation (int or Tuple[int, int]): the spacing between kernel elements. Default: 1
+        mask (Tensor[batch_size, offset_groups * kernel_height * kernel_width,
+            out_height, out_width]): masks to be applied for each position in the
+            convolution kernel.
```
For a follow-up PR, it might be good to mention that now `DeformConv2d` implements modulation from https://arxiv.org/abs/1811.11168 as well (also known as DeformConv v2)

fmassa reviewed (approved):
Thanks a lot for the PR!

I'm accepting the `config.yml` changes for now to get this PR merged, but there might be better ways of handling the caching and CI.

I'll create an issue to track those down.
